File: packages/sdk-typescript/src/card.ts
// SPDX-License-Identifier: Apache-2.0
import type { AgentCard, PublicKey, SkillDef, RateLimit } from './types.js'

interface BuildCardOptions {
  name: string
  version: string
  description: string
  url: string
  specializations: string[]
  models: Array<{ provider: string; model: string }>
  skills: SkillDef[]
  publicKeys: PublicKey[]
  rateLimit: RateLimit
  cardTTL: number
}

export function buildAgentCard(opts: BuildCardOptions): AgentCard {
  const domain = new URL(opts.url).hostname
  return {
    id: `agent://${domain}`,
    name: opts.name,
    version: opts.version,
    description: opts.description,
    url: opts.url,
    protocolVersion: '1.1',
    specializations: opts.specializations,
    models: opts.models,
    skills: opts.skills,
    publicKeys: opts.publicKeys,
    auth: { schemes: ['bearer', 'none'] },
    rateLimit: opts.rateLimit,
    cardTTL: opts.cardTTL,
    endpoints: {
      intro: '/agent/intro',
      message: '/agent/message',
      task: '/agent/task',
      taskStatus: '/agent/task/:taskId',
      stream: '/agent/stream',
      health: '/agent/health',
    },
  }
}

File: packages/sdk-python/src/samvad/card.py
# SPDX-License-Identifier: Apache-2.0
from urllib.parse import urlparse

from .types import AgentCard, AgentCardAuth, AgentCardEndpoints, AgentCardModel, PublicKey, RateLimit, SkillDef


def build_agent_card(
    *,
    agent_id: str | None = None,
    name: str,
    version: str,
    description: str,
    url: str,
    specializations: list[str],
    models: list[dict],
    skills: list[SkillDef],
    public_keys: list[PublicKey],
    rate_limit: RateLimit,
    card_ttl: int = 300,
) -> AgentCard:
    if not agent_id:
        domain = urlparse(url).hostname or "unknown"
        agent_id = f"agent://{domain}"

    has_authenticated_skill = any(s.trust == "authenticated" for s in skills)
    auth_schemes = ["bearer", "none"] if has_authenticated_skill else ["none"]
    auth = AgentCardAuth(schemes=auth_schemes)

    base = url.rstrip("/")
    return AgentCard(
        id=agent_id,
        name=name,
        version=version,
        description=description,
        url=url,
        protocolVersion="1.2",
        specializations=specializations,
        models=[AgentCardModel(**m) if isinstance(m, dict) else m for m in models],
        skills=skills,
        publicKeys=public_keys,
        auth=auth,
        rateLimit=rate_limit,
        cardTTL=card_ttl,
        endpoints=AgentCardEndpoints(
            intro=f"{base}/agent/intro",
            message=f"{base}/agent/message",
            task=f"{base}/agent/task",
            taskStatus=f"{base}/agent/task/{{taskId}}",
            stream=f"{base}/agent/stream",
            health=f"{base}/agent/health",
        ),
    )
